### PR TITLE
feat: Record summary of execution times of support bundle operations (collect/redact/analyse)

### LINKED ISSUE #936: Design proposal for adding instrumentation to the Troubleshoot code

### PR DESCRIPTION
**Pull request by banjoh**
## Description, Motivation and Context

This PR contains an implementation of collecting traces using [OTEL library](https://github.com/open-telemetry/opentelemetry-go). The traces are then used to record a summary of an execution of `support-bundle` and `preflight` commands. The summary is printed to stdout if the `--debug` flag is passed. When a support bundle is collected, the execution summary is stored in `execution-data/summary.txt` location.

Related to #923

Example output on a k8s cluster

**Support bundle**
```sh
[evans] $ ./bin/support-bundle examples/support-bundle/sample-supportbundle.yaml --debug

support-bundle-2023-02-03T19_02_28.tar.gz

========= Collectors summary ==========
run-pod/static-hi                : 2,078ms
cluster-resources                : 1,833ms
logs/all-logs                    : 93ms
Host OS Info                     : 23ms
logs/app=example,component=nginx : 4ms
cluster-info                     : 3ms

========= Redactors summary ==========
In-cluster collectors : 514ms
Host collectors       : 1ms

========= Analysers summary ==========
Custom resource definition rook                         : 2ms
Container Runtime                                       : 0ms
Must have at least 3 nodes in the cluster               : 0ms
Total CPU Cores in the cluster is 4 or greater          : 0ms
Each node must have at least 40 GB of ephemeral storage : 0ms
Required storage classes                                : 0ms
Ingress connect-to-me                                   : 0ms
Required Kubernetes Version                             : 0ms
Said hi!                                                : 0ms
api/api Deployment Status                               : 0ms

Duration: 4,594ms
```

**Preflights**
```sh
[evans] $ ./bin/preflight examples/preflight/sample-preflight.yaml --debug

========= Collectors summary ==========
cluster-resources : 1,826ms
cluster-info      : 2ms

========= Redactors summary ==========
No redactors executed

========= Analysers summary ==========
Ingress                                                                                          : 3ms
Kubernetes Distribution                                                                          : 0ms
Container Runtime                                                                                : 0ms
Must have at least 3 nodes in the cluster, with 5 recommended                                    : 0ms
Every node in the cluster must have at least 8 GB of memory, with 32 GB recommended              : 0ms
Every node in the cluster must have at least 40 GB of ephemeral storage, with 100 GB recommended : 0ms
Total CPU Cores in the cluster is 4 or greater                                                   : 0ms
Required storage classes                                                                         : 0ms
Required Kubernetes Version                                                                      : 0ms

Duration: 3,095ms
```

## Checklist

- [x] New and existing tests pass locally with introduced changes.
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] The commit message(s) are informative and highlight any breaking changes
- [ ] Any documentation required has been added/updated. For changes to https://troubleshoot.sh/ create a PR [here](https://github.com/replicatedhq/troubleshoot.sh/pulls)

## Does this PR introduce a breaking change?
- [ ] Yes
- [x] No

### CODE REVIEW

**Comment by CpuID** [comment]
At a high level, this feels like it makes sense to use something like OpenTelemetry.

Even though you mentioned (in the standup) that it's more commonly used in distributed systems, it's still just data that forms a Gantt chart at the end of the day, which if there are multiple background tasks running in parallel (collectors etc) being able to visualise the spans together as a trace feels logical.

An upstream doc for anyone who hasn't done any distributed tracing previously: https://opentelemetry.io/docs/concepts/signals/traces/

**Comment by DexterYan** [comment]
Thank you @banjoh. I am thinking when we enable this tracing mode, should we also get a summary data of current cluster, like how many nodes, pods. namespaces are running and store it with tracing data. So after collecting both tracing data and cluster data, we can get more details to compare the performance in different kinds of clusters.

**Comment by xavpaice** [comment]
Regards the summary data, I would prefer that to be a separate task and made available for all collection runs - see #731

**Comment by diamonwiggins** [comment]
Is there another CLI out there we could reference as a good example of using something like OpenTelemetry when everything is client side?

**Comment by banjoh** [comment]
> Is there another CLI out there we could reference as a good example of using something like OpenTelemetry when everything is client side?

I've not found one so far besides sample implementations such as [this](https://github.com/open-telemetry/opentelemetry-go/tree/main/example/fib), but I'll keep looking. I'm sure there is something out there

File: pkg/supportbundle/collect.go [hunk 1]
```diff
@@ -140,17 +160,21 @@ func runCollectors(collectors []*troubleshootv1beta2.Collect, additionalRedactor
 			if _, ok := collector.(*collect.CollectClusterResources); !ok {
 				msg := fmt.Sprintf("skipping collector %s with insufficient RBAC permissions", collector.Title())
 				opts.CollectorProgressCallback(opts.ProgressChan, msg)
+				span.End()
```
Comment (emosbaugh) [inline pkg/supportbundle/collect.go:167]: i think you'll want to indicate that it was excluded here in the span

File: pkg/preflight/collect.go [hunk 2]
```diff
@@ -217,6 +250,7 @@ func Collect(opts CollectOpts, p *troubleshootv1beta2.Preflight) (CollectResult,
 					TotalCount:     len(allCollectors),
 					Collectors:     collectorList,
 				}
+				span.End()
```
Comment (emosbaugh) [inline pkg/preflight/collect.go:257]: i think you'll want to indicate that it was excluded here in the span

**Comment by CpuID** [comment]
Loving how the output of this looks right now :) Makes it useful standalone, awesome.

Haven't looked closely at the code under the hood yet though.

**Review by emosbaugh — COMMENTED** [review]
This makes sense to me. My understanding is there is little to no overhead if tracing is not configured.

**Comment by banjoh** [comment]
> My understanding is there is little to no overhead if tracing is not configured.

That is correct. Collecting operation time spans will effectively be a no-op if there is no trace provider + exporter configured. The spans are collected but dropped when exporting them.

**Review by diamonwiggins — DISMISSED** [review]

**Comment by banjoh** [comment]
I added succeeded/failed/excluded status to the execution summary. Now the status looks like so

```sh
============ Collectors summary =============
Suceeded (S), eXcluded (X), Failed (F)
=============================================
cluster-resources (S)                : 1,821ms
logs/all-logs (S)                    : 63ms
Host OS Info (S)                     : 24ms
cluster-info (S)                     : 1ms
logs/app=example,component=nginx (X) : 0ms
run-pod/static-hi (X)                : 0ms

============ Redactors summary =============
In-cluster collectors : 483ms
Host collectors       : 0ms

============= Analyzers summary =============
Suceeded (S), eXcluded (X), Failed (F)
=============================================
Custom resource definition rook (S)                         : 2ms
Container Runtime (S)                                       : 0ms
Each node must have at least 40 GB of ephemeral storage (S) : 0ms
Must have at least 3 nodes in the cluster (S)               : 0ms
Required storage classes (S)                                : 0ms
Required Kubernetes Version (S)                             : 0ms
Said hi! (S)                                                : 0ms
Total CPU Cores in the cluster is 4 or greater (X)          : 0ms
Ingress connect-to-me (X)                                   : 0ms
api/api Deployment Status (X)                               : 0ms

Duration: 2,422ms
```

**Review by diamonwiggins — APPROVED** [review]

**Pull request merged**